### PR TITLE
add all gene panels to mip analysis

### PR DESCRIPTION
**Pull request by peterpru**
## Description
In the latest PR, gene panel master list was changed to a constant class, creating a bug where only the gene panels selected by the customer are added to the analysis. Before this, all gene panels were added, with the panels selected by the customer being displayed as default.

Example for MIP-DNA case inheron: gene panels selected by customer were Cardiology, and as this panel is part of the master list, all panels should have been added into the analysis. After the last PR, the MIP case inheron only had Cardiology and OMIM-AUTO used in the MIP analysis, as seen in screenshot below for the gene_panels.bed file:
image

### Added

-

### Changed

-

### Fixed

- Add all gene panels when the default gene panel is found in the gene panel master list.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b fix-missing-genepanels-in-scout -a
    ```

### How to test

- [ ] Start dry run of MIP case inheron: `cg workflow mip-dna start --mip-dry-run inheron`

### Expected test outcome

- [ ] Check that the gene_panels.bed file has all gene panels that are included in the masterlist.
- [ ] Take a screenshot and attach or copy/paste the output.
image

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Review by henrikstranneheim — COMMENTED** [review]
Good work! How about adding unit tests for these 2 functions to make sure they do what we think and avoid surprises?

**Comment by peterpru** [comment]
> Good work! How about adding unit tests for these 2 functions to make sure they do what we think and avoid surprises?

Good idea, I have a new appreciation for unit tests, I will add these here.

Question for @dnil: Gene panel ET (Endocrine Tumors) is currently not in the cg masterlist of gene panels. Should it be? In the cases that ET is used now, it would only add OMIM-AUTO and not add any other panels. See below for an example:

Test output example from hasta stage:
Case inheron has as default_panels CARDIOLOGY, and all panels are added:
image
image

Case validcolt has as default panels ET:
image
image

**Comment by dnil** [comment]
@peterpru good question; it is kept up to date and active. If they are used to ordering that way I guess its fine, but I can double check with the folks that usually order it!

**Comment by dnil** [comment]
Yeah, it's intentional. And possibly a little misguided, as it is likely to keep those genes clear of general analysis, while all of OMIM is included anyway. Thanks for the attentiveness though!

**Comment by peterpru** [comment]
@dnil Thank you very much for checking this!

File: tests/meta/workflow/test_analysis.py [hunk 1]
```diff
@@ -29,3 +31,25 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
 
     # THEN the expected slurm QOS should be returned
     assert slurm_qos is expected_slurm_qos
+
+
+def test_gene_panels_correctly_added():
+    """Test get correct gene panel list"""
+
+    # GIVEN a case that has a gene panel included in the master list:
```
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:39]:
```suggestion
    # GIVEN a case that has a gene panel included in the gene panel master list:
```

File: tests/meta/workflow/test_analysis.py [hunk 2]
```diff
@@ -29,3 +31,25 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    default_panels_included: List[str] = [GenePanelMasterList.get_panel_names()[0]]
+    master_list = GenePanelMasterList.get_panel_names()
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
```
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:42]:
```suggestion
    master_list: List[str] = GenePanelMasterList.get_panel_names()

    # WHEN converting the gene panels between the default and the gene_panel_master_list:
```

File: tests/meta/workflow/test_analysis.py [hunk 3]
```diff
@@ -29,3 +31,25 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    assert set(list_of_gene_panels_used) == set(master_list)
+
+    # GIVEN a case that has a gene panel that is NOT included in the master list:
+    default_panels_not_included: List[str] = ["ANOTHER_PANEL"]
+
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
+    list_of_gene_panels_used = MipAnalysisAPI.convert_panels("cust002", default_panels_not_included)
+
+    # THEN the list_of_gene_panels_used should return the custom panel and OMIM-AUTO:
+    assert set(list_of_gene_panels_used) == set(default_panels_not_included + ["OMIM-AUTO"])
```
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:55]: This should be a separate test

File: tests/meta/workflow/test_analysis.py [hunk 4]
```diff
@@ -29,3 +31,25 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    default_panels_included: List[str] = [GenePanelMasterList.get_panel_names()[0]]
+    master_list = GenePanelMasterList.get_panel_names()
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
+    list_of_gene_panels_used = MipAnalysisAPI.convert_panels("cust002", default_panels_included)
+
+    # THEN the list_of_gene_panels_used should return all gene panels:
+    assert set(list_of_gene_panels_used) == set(master_list)
+
+    # GIVEN a case that has a gene panel that is NOT included in the master list:
+    default_panels_not_included: List[str] = ["ANOTHER_PANEL"]
```
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:49]:
```suggestion
    default_panels_not_included: List[str] = ["PANEL_NOT_IN_GENE_PANEL_MASTER_LIST"]
```

File: tests/meta/workflow/test_analysis.py [hunk 5]
```diff
@@ -29,3 +31,25 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
+    list_of_gene_panels_used = MipAnalysisAPI.convert_panels("cust002", default_panels_not_included)
```
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:52]: Do we have a cust test fixture that you could reuse?

File: tests/meta/workflow/test_analysis.py (hunk 3 above)
Comment (henrikstranneheim) [inline tests/meta/workflow/test_analysis.py:55]: I think you should use the new shiny constant class you just created instead of hard-coding `OMIM_AUTO`

**Review by henrikstranneheim — COMMENTED** [review]
Nice with tests 👍

File: tests/meta/workflow/test_analysis.py (hunk 5 above)
Comment (peterpru) [reply tests/meta/workflow/test_analysis.py]: Yes, added the cust fixture, and divided the test up into two separate tests.

File: tests/meta/workflow/test_analysis.py [hunk 6]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
 
     # THEN the expected slurm QOS should be returned
     assert slurm_qos is expected_slurm_qos
+
+
+def test_gene_panels_correctly_added(customer_id):
+    """Test get correct gene panel list"""
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:37]:
```suggestion
    """Test get correct gene panel list."""
```

File: tests/meta/workflow/test_analysis.py [hunk 7]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+
+    # GIVEN a case that has a gene panel included in the gene panel master list:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:39]:
```suggestion
    # GIVEN a case that has a gene panel included in the gene panel master list
```

File: tests/meta/workflow/test_analysis.py [hunk 8]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    default_panels_included: List[str] = [GenePanelMasterList.get_panel_names()[0]]
+    master_list: List[str] = GenePanelMasterList.get_panel_names()
+
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:43]:
```suggestion
    # WHEN converting the gene panels between the default and the gene_panel_master_list
```

File: tests/meta/workflow/test_analysis.py [hunk 9]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    list_of_gene_panels_used = MipAnalysisAPI.convert_panels(
+        customer=customer_id, default_panels=default_panels_included
+    )
+
+    # THEN the list_of_gene_panels_used should return all gene panels:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:48]:
```suggestion
    # THEN the list_of_gene_panels_used should return all gene panels
```

File: tests/meta/workflow/test_analysis.py [hunk 10]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    assert set(list_of_gene_panels_used) == set(master_list)
+
+
+def test_gene_panels_not_added(customer_id):
+    """Test get only OMIM-AUTO and custom gene panel list"""
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:53]:
```suggestion
    """Test get only OMIM-AUTO and custom gene panel list."""
```

File: tests/meta/workflow/test_analysis.py [hunk 11]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    # GIVEN a case that has a gene panel that is NOT included in the gene panel master list:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:54]:
```suggestion
    # GIVEN a case that has a gene panel that is NOT included in the gene panel master list
```

File: tests/meta/workflow/test_analysis.py [hunk 12]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    default_panels_not_included: List[str] = ["PANEL_NOT_IN_GENE_PANEL_MASTER_LIST"]
+
+    # WHEN converting the gene panels between the default and the gene_panel_master_list:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:57]:
```suggestion
    # WHEN converting the gene panels between the default and the gene_panel_master_list
```

File: tests/meta/workflow/test_analysis.py [hunk 13]
```diff
@@ -29,3 +31,35 @@ def test_get_slurm_qos_for_case(mocker, case_id: str, priority, expected_slurm_q
...
+    list_of_gene_panels_used = MipAnalysisAPI.convert_panels(
+        customer=customer_id, default_panels=default_panels_not_included
+    )
+
+    # THEN the list_of_gene_panels_used should return the custom panel and OMIM-AUTO:
```
Comment (ivadym) [inline tests/meta/workflow/test_analysis.py:62]:
```suggestion
    # THEN the list_of_gene_panels_used should return the custom panel and OMIM-AUTO
```

**Review by ivadym — APPROVED** [review]
Well done 💯 

I've just added some minor formatting comments

**Comment by peterpru** [comment]
Thank you @henrikstranneheim and @ivadym for your reviews.

Deployed successfully in production.
image

**Comment by peterpru** [comment]
Sanity check: this case `gentleram` was started by the crontab automation for mip-dna after deployment, and includes all gene panels from the master list in the analysis.
image

**Pull request merged**